Add wp_prime_network_option_caches function in option.php

spacedmonkey opened the pull request:
Introduced the new function wp_prime_network_option_caches in option.php. This function simplifies the handling of network option caches, enhancing the speed of operations. In the wp_load_core_site_options function, removed cache processing code and replaced it with a call to the new function. These changes improve code efficiency and readability.

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket:  https://core.trac.wordpress.org/ticket/61053

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

jonnynews committed:
Add wp_prime_network_option_caches function in option.php

Introduced the new function wp_prime_network_option_caches in option.php. This function simplifies the handling of network option caches, enhancing the speed of operations. In the wp_load_core_site_options function, removed cache processing code and replaced it with a call to the new function. These changes improve code efficiency and readability.

jonnynews committed:
Merge branch 'trunk' into try/prime_network_options

spacedmonkey committed:
Improve option caching in WP site options

The commit adds more sophisticated logic in the site options cache. It not only fetches and caches the options, but also identifies and caches those options that were not found ('notoptions'). Only when 'notoptions' cache has been modified, does it update this cache, thus optimizing efficiency.

spacedmonkey committed:
Fallback to site options if not multisite.

spacedmonkey committed:
Return logic.

spacedmonkey committed:
Refactor caching logic in wp_prime_network_option_caches

This commit simplifies the caching logic in the wp_prime_network_option_caches function. It uses the cache_keys array as a map instead of a simple list, allowing us to avoid duplicate attempts to generate cache keys. A $cache_group variable is also introduced to replace hard-coded 'site-options' string, enhancing code readability and maintenance.

spacedmonkey committed:
Merge branch 'trunk' into try/prime_network_options

spacedmonkey committed:
Sort post types for cache key generation

The code has been updated to sort post types before the cache key generation to ensure consistency. Also, a default orderby value of 'date' has been added for the same purpose. Changes have been tested by adding a new test case to verify the query cache for differently ordered post types.

spacedmonkey committed:
Add unit tests for wp_prime_network_option_caches method

This commit adds several unit tests for the wp_prime_network_option_caches method in WordPress. Specifically, it includes tests for priming multiple options, handling a mix of primed and unprimed options, and tests related to prime options on a different network.

spacedmonkey committed:
Add 'WPLANG' to core options list

This commit introduces 'WPLANG' to the array of core options in the 'option.php' file. This ensures that the language setting for Wordpress is included in the list of options that are cached for network use.

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -247,6 +247,95 @@ function get_option( $option, $default_value = false ) {
+		return;
+	}
+
+	$core_options_in = "'" . implode( "', '", $options_to_prime ) . "'";
```
```suggestion
	$core_options_in = "'" . implode( "', '", esc_sql( $options_to_prime ) ) . "'";
```

Another approach may be to use the `array_fill()` approach in `wp_prime_option_caches()`

Option names can contain quotes & other meaningful characters. Wild:

```
wp> update_site_option( "Can Option Names Contain 'quotes'", "yes" );
=> phar:///usr/local/src/wp-cli/bin/wp/vendor/wp-cli/shell-command/src/WP_CLI/Shell/REPL.php:52:
bool(true)
wp> get_site_option( "Can Option Names Contain 'quotes'" );
=> phar:///usr/local/src/wp-cli/bin/wp/vendor/wp-cli/shell-command/src/WP_CLI/Shell/REPL.php:52:
string(3) "yes"
wp> update_site_option( 'Can Option Names Contain "quotes"', "yes" );
=> phar:///usr/local/src/wp-cli/bin/wp/vendor/wp-cli/shell-command/src/WP_CLI/Shell/REPL.php:52:
bool(true)
wp> get_site_option( 'Can Option Names Contain "quotes"' );
=> phar:///usr/local/src/wp-cli/bin/wp/vendor/wp-cli/shell-command/src/WP_CLI/Shell/REPL.php:52:
string(3) "yes"
```

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -247,6 +247,95 @@ function get_option( $option, $default_value = false ) {
+	}
+
+	$core_options_in = "'" . implode( "', '", $options_to_prime ) . "'";
+	$options         = $wpdb->get_results( $wpdb->prepare( "SELECT meta_key, meta_value FROM $wpdb->sitemeta WHERE meta_key IN ($core_options_in) AND site_id = %d", $network_id ) );
```
This overrides the passed value `$options`. Best to use a different name so it's easier to maintain in to the future.

:1234: This will need to be renamed where it's used later.

```suggestion
	$results         = $wpdb->get_results( $wpdb->prepare( "SELECT meta_key, meta_value FROM $wpdb->sitemeta WHERE meta_key IN ($core_options_in) AND site_id = %d", $network_id ) );
```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,161 @@
+		// Store the initial database query count.
+		$initial_query_count = get_num_queries();
+
+		// Check that options are only in the 'options' cache group.
```
I'm not sure what this comment means.

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,161 @@
+			'Additional database queries were made.'
+		);
+	}
+
+
```
```suggestion

```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,161 @@
+		 */
+		foreach ( $options_to_prime as $option ) {
+			update_network_option( null, $option, "value_$option" );
+		}
```
`update_network_option()` primes the cache so the getter calls below won't trigger DB queries if the priming function fails.

You'll need to do something similar to the test for `wp_prime_option_caches()`

https://github.com/WordPress/wordpress-develop/blob/17a01ed185f38f87362eb7588908dc95ec6cbe23/tests/phpunit/tests/option/wpPrimeOptionCaches.php#L24-L33

:1234: This applies to the other tests too.

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -247,6 +247,95 @@ function get_option( $option, $default_value = false ) {
+ *
+ * @param string[] $options An array of option names to be loaded.
+ */
+function wp_prime_network_option_caches( $network_id = null, array $options = array() ) {
```
Do you think it's worth having `wp_prime_site_option_caches( array $options = [] )` as a wrapper for consistency with the other option functions?

mukeshpanchal27 replied:
The documentation is missing for new param.

peterwilsoncc reviewed (commented):
I've added a few notes inline. This seems to be testing well but I think the code needs a little tidying up.

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -2415,7 +2475,9 @@ function get_site_transient( $transient ) {
 		$transient_option = '_site_transient_' . $transient;
 		if ( ! in_array( $transient, $no_timeout, true ) ) {
 			$transient_timeout = '_site_transient_timeout_' . $transient;
-			$timeout           = get_site_option( $transient_timeout );
+			wp_prime_network_option_caches( null, array( $transient_option, $transient_timeout ) );
+
+			$timeout = get_site_option( $transient_timeout );
```
Same will be needed in `set_site_transient()`.

jonnynews replied:
Fixed.

spacedmonkey committed:
Update wpPrimeNetworkOptionCaches tests to handle multisite support

Added multisite support to wpPrimeNetworkOptionCaches tests by separately setting up cache groups for multisite and single site. Also included additional checks for option caching in multisite setup.

spacedmonkey committed:
Feedback.

spacedmonkey committed:
Refactor cache key generation and deletion in tests

The cache key generation and deletion stages have been refactored in the test scripts. Instead of repeating the same logic to generate cache keys, a separate code block defines all the cache keys up front. This change simplifies cache deletion and read operations, making the test scripts cleaner and easier to understand.

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,306 @@
+			);
+		}
+
+		$this->assertFalse( get_network_option( $network_id, 'option404notfound' ), "$option should return false as option does not exist" );
```
```suggestion
		$this->assertFalse( get_network_option( $network_id, 'option404notfound' ), 'option404notfound should return false as option does not exist' );
```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,306 @@
+				'domain' => 'wordpress.org',
+				'path'   => '/',
+			)
+		);
```
This can be moved to a shared fixture as it's created twice and is quite resource intensive.

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,306 @@
+			get_num_queries(),
+			'Additional database queries were made.'
+		);
+	}
+	/**
```
I suspect GH got confused with the empty lines.

```suggestion
	}

	/**
```

peterwilsoncc reviewed (commented):
Source changes look good to me, a couple more notes in the tests.

Otherwise, the only question that remains is whether we should also add `wp_prime_site_option_caches` as a wrapper similar to all the other site/network option functions.

My inclination is to do so.

spacedmonkey committed:
Feedback.

mukeshpanchal27 commented on src/wp-includes/option.php:
```diff
@@ -247,6 +247,105 @@ function get_option( $option, $default_value = false ) {
 }
 
+/**
+ *
```
Function Summary and Description is missing.

jonnynews replied:
Fixed.

spacedmonkey committed:
Merge branch 'trunk' into try/prime_network_options

spacedmonkey committed:
Add comment.

spacedmonkey committed:
Merge branch 'trunk' into try/prime_network_options

spacedmonkey committed:
Refactor option caching functions in WordPress

This commit refactors the option caching functions in the WordPress source code for efficiency. It reorders the functions and renames wp_prime_network_option_caches to wp_prime_site_option_caches to enhance clarity. The update ensures optimal caching of network options with a single database query, which significantly increases performance.

spacedmonkey committed:
Add network_id parameter to wp_prime_network_option_caches

This commit adds an optional network_id parameter to the wp_prime_network_option_caches function in the options.php file. This offers the flexibility of specifying a network ID when priming network option caches, defaulting to the current network ID if none is provided.

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,143 @@ function wp_load_alloptions( $force_cache = false ) {
 
 /**
- * Loads and primes caches of certain often requested network options if is_multisite().
+ * Primes specific network options into the cache with a single database query.
```
```suggestion
 * Primes specific network options for the current network into the cache with a single database query.
```

Similar to the difference between `get_site_option` and `get_network_option`.

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,143 @@ function wp_load_alloptions( $force_cache = false ) {
  *
- * @param int $network_id Optional. Network ID of network for which to prime network options cache. Defaults to current network.
+ * @param int $network_id ID of the network. Can be null to default to the current network ID.
+ * @param string[] $options An array of option names to be loaded.
```
```suggestion
 * @param int      $network_id ID of the network. Can be null to default to the current network ID.
 * @param string[] $options    An array of option names to be loaded.
```

WPCS

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,310 @@
+	}
+
+	/**
+	 * Tests that wp_prime_network_option_caches() is run twice
```
```suggestion
	 * Tests that running wp_prime_network_option_caches() twice does not requery options.
```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,310 @@
+			$this->assertFalse( wp_cache_get( $cache_key, $cache_group ), "$option was not deleted from the cache." );
+		}
+
+		// Call the wp_prime_option_caches function to prime the options.
```
```suggestion
		// Call the wp_prime_network_option_caches() function to prime the options.
```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,310 @@
+			'option3',
+		);
+
+		// Call the wp_prime_option_caches function to prime the options.
```
```suggestion
		// Call the wp_prime_network_option_caches() function to prime the options.
```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,310 @@
+		// Store the initial database query count.
+		$initial_query_count = get_num_queries();
+
+		// Call the wp_prime_option_caches function second time
```
```suggestion
		// Call the wp_prime_network_option_caches() function second time
```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,310 @@
+			$this->assertFalse( wp_cache_get( $cache_key, $cache_group ), "$option was not deleted from the cache." );
+		}
+
+		// Call the wp_prime_option_caches function to prime the options.
```
```suggestion
		// Call the wp_prime_network_option_caches() function to prime the options.
```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,310 @@
+			$this->assertFalse( wp_cache_get( $cache_key, $cache_group ), "$option was not deleted from the cache." );
+		}
+
+		// Call the wp_prime_option_caches function to prime the options.
```
```suggestion
		// Call the wp_prime_network_option_caches() function to prime the options.
```

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,143 @@ function wp_load_alloptions( $force_cache = false ) {
+ *
+ * @since 3.0.0
+ * @since 6.3.0 Also prime caches for network options when persistent object cache is enabled.
+ * @since 6.6.0 Uses wp_prime_network_option_caches.
```
```suggestion
 * @since 6.6.0 Uses wp_prime_network_option_caches().
```

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,143 @@ function wp_load_alloptions( $force_cache = false ) {
+ *
+ * Only network options that do not already exist in cache will be loaded.
+ *
+ * If site is not multisite, then call wp_prime_option_caches.
```
```suggestion
 * If site is not multisite, then call wp_prime_option_caches().
```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,310 @@
+	 */
+	public function test_wp_prime_network_option_caches_multiple_networks() {
+		$network_id = get_current_network_id();
+		if ( is_multisite() ) {
+			$cache_group = 'site-options';
+		} else {
+			$cache_group = 'options';
+		}
```
In ms-required group.

```suggestion
		$cache_group = 'site-options';
```

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,143 @@ function wp_load_alloptions( $force_cache = false ) {
- * @since 6.3.0 Also prime caches for network options when persistent object cache is enabled.
+ * Only network options that do not already exist in cache will be loaded.
+ *
+ * If site is not multisite, then call wp_prime_option_caches.
```
```suggestion
 * If site is not multisite, then call wp_prime_option_caches().
```

peterwilsoncc commented on tests/phpunit/tests/option/wpPrimeNetworkOptionCaches.php:
```diff
@@ -0,0 +1,310 @@
+
+		$cache_keys = array();
+		foreach ( $options_to_prime as $option ) {
+			if ( is_multisite() ) {
+				$cache_key = "$network_id:$option";
+			} else {
+				$cache_key = $option;
+			}
```
ms-required group.

```suggestion
			$cache_key = "$network_id:$option";
```

peterwilsoncc reviewed (commented):
A few more things that need tidying up before it's commit ready.

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

spacedmonkey committed:
Refactor variable assignment alignment

In the PHPUnit test file for priming network options, the alignment of the equals (=) operator in variable assignments has been made consistent. This applies to the declarations of `$network_id` and `$cache_key`, improving readability and adherence to code style guidelines.

spacedmonkey commented:
Amazing feedback @peterwilsoncc . It is really appricated. Your best code review! Legend.

joemcgill commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,143 @@ function wp_load_alloptions( $force_cache = false ) {
 		return;
 	}
 
 	if ( empty( $network_id ) ) {
```
Nothing is enforcing that $network_id is an `int` in this function even though that's how it's documented. I'd suggest adding a safety check here, like:

```suggestion
	if ( empty( $network_id ) || ! is_int( $network_id ) ) {
```

Or even consider returning early if an invalid value is passed.

spacedmonkey replied:
Good catch. Here is a fix. https://github.com/WordPress/wordpress-develop/pull/6311/commits/482cc816ea9f970611f8058b2a071cefc3cfe337 

Logic is copied from 

https://github.com/WordPress/wordpress-develop/blob/2a21dd9fef93bcb0405d9f189286402b972faabe/src/wp-includes/option.php#L1856-L1865

joemcgill commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,143 @@ function wp_load_alloptions( $force_cache = false ) {
  * @global wpdb $wpdb WordPress database abstraction object.
  *
- * @param int $network_id Optional. Network ID of network for which to prime network options cache. Defaults to current network.
+ * @param int      $network_id ID of the network. Can be null to default to the current network ID.
```
Nit: consider either changing the default to `0` or documenting as `int|null`.

Alternately, this function could make the network ID a required param, like the other `*_network_options()` functions, and move the `get_current_network_id()` call to the `wp_prime_site_option_caches()` wrapper.

spacedmonkey replied:
Fixed. Keeping the logic the same as get_network_option / get_site_option.

joemcgill reviewed (commented):
Couple of notes. Great work here!

spacedmonkey committed:
Update wp_prime_network_option_caches() function parameters

The function wp_prime_network_option_caches() was updated to no longer allow a null value for the $network_id parameter. A validity check for the $network_id parameter was also added. If the parameter is not a number, the function will return early.

spacedmonkey commented:
Thanks for the feedback @joemcgill . Actioned feedback.

joemcgill commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,150 @@ function wp_load_alloptions( $force_cache = false ) {
+ * @param string[] $options An array of option names to be loaded.
+ */
+function wp_prime_site_option_caches( array $options = array() ) {
+	wp_prime_network_option_caches( null, $options );
```
The updates in https://github.com/WordPress/wordpress-develop/commit/482cc816ea9f970611f8058b2a071cefc3cfe337 are a nice change, but I would also update the call to `wp_prime_network_option_caches()` here so it doesn't pass a `null` value.

```suggestion
	wp_prime_network_option_caches( 0, $options );
```

spacedmonkey replied:
@joemcgill @mukeshpanchal27 Passing null is what is done in functions like `get_network_option`. Why pass null there and 0 here?

joemcgill replied:
Interesting. I'd argue that `get_site_option()` (which is the only place where I see `null` being passed to `get_network_option()`) should also be passing `0`, but I think the consistency between the two sets of functions makes sense.

joemcgill commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,150 @@ function wp_load_alloptions( $force_cache = false ) {
  * @global wpdb $wpdb WordPress database abstraction object.
  *
- * @param int $network_id Optional. Network ID of network for which to prime network options cache. Defaults to current network.
+ * @param int      $network_id ID of the network. Can be null to default to the current network ID.
```
Related to my previous suggestion. If we're enforcing `int` values, Let's update the docs as well.
```suggestion
 * @param int      $network_id ID of the network. Passing 0 will default to the current network ID.
```

joemcgill reviewed (approved):
Thanks, that looks pretty good. One more small suggestion, but pre-approving since you addressed my main concern.

mukeshpanchal27 commented on src/wp-includes/option.php:
```diff
@@ -643,50 +643,150 @@ function wp_load_alloptions( $force_cache = false ) {
+ *
+ * @param string[] $options An array of option names to be loaded.
+ */
+function wp_prime_site_option_caches( array $options = array() ) {
```
```suggestion
function wp_prime_site_option_caches( array $options ) {
```

Similar to `wp_prime_network_option_caches` ?

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

joemcgill reviewed (approved):
Left feedback inline. Approving since using `null` is consistent with how `get_site_option()` was previously impelmented.

spacedmonkey committed:
Merge branch 'trunk' into try/prime_network_options

spacedmonkey commented:
Committed https://github.com/WordPress/wordpress-develop/commit/6662cb0bef6bdd3fd2aa02db92f5781a4b5db657
